locale/en/docs/guides/dont-block-the-event-loop.md

```diff
@@ -0,0 +1,435 @@
+---
+title: Don't Block the Event Loop (or the Worker Pool)
+layout: docs.hbs
+---
+
+# Don't Block the Event Loop (or the Worker Pool)
+
+## Should you read this guide?
+If you're writing a server using Node, keep reading.
+If you're writing command-line scripts with Node, this guide might still come in handy, but it isn't crucial.
+
+## TL; DR
+Node.js runs JavaScript code in the Event Loop and handles I/O in a Worker Pool (the threadpool).
+Node scales well, sometimes better than more heavyweight approaches like Apache.
+The secret to Node's scalability is that it uses a small number of threads to handle many clients.
+If Node can make do with fewer threads, then it can spend more of your system's time and memory working on clients rather than on paying overheads for threads.
+
+Here's a good rule of thumb for keeping your Node server speedy:
+*Node is only fast when the work associated with each client at any given time is "small"*.
+
+This applies to work both on the Event Loop and on the Worker Pool.
+
+## Why should I avoid blocking the Event Loop and the Worker Pool?
+Node uses a small number of threads to handle many clients.
+While a thread is working on behalf of one client, it can't work for any other clients.
+
+This provides two motivations for blocking neither the Event Loop nor the Worker Pool.
+Since both the Event Loop and the Workers respond to events, we will refer to these threads generally as *Event Handlers* or *Event-Handler Threads*.
+All threads in Node are Event Handlers: one Event Loop and several Workers.
+
+1. Performance: If you regularly perform heavyweight tasks on an Event Handler, the *throughput* (requests/second) of your server will suffer.
+2. Security: If it is possible that for certain input you might block an Event Handler, a malicious client might be able to figure out the "evil input", make your code block, and keep you from working on other clients. This would be a Denial of Service attack.
+
+## A quick review of Node's architecture
+
+Node uses the Event-Driven Architecture: it has an Event Loop for orchestration and a Worker Pool for expensive tasks.
+
+## Queues and Cooperative Multitasking
+Both the Event Loop and the Worker Pool use *queues*, which are data structures with one or more producers who place items into the queue, and one or more consumers who remove items from the queue.
+
+The Event Loop's queue is full of pending events.
+When the Event Loop sees a new event, it invokes its callback, which is JavaScript code that sees your Node application.
+The Event Loop is the only consumer of this queue.
+The Node runtime produces events for this queue.
+Specifically, Node asks libuv to monitor file descriptors using [epoll](http://man7.org/linux/man-pages/man7/epoll.7.html)), and the Event Loop's queue consists of the ready descriptors.
```

Comment:
only on linux, its kqueue on OS X, and other mechanisms on Windows

Reply:
Yes, good catch.

Reply:
What do you think about saying "an epoll-like system call" with a link to the epoll man page? Part of my goal is to point users to resources that will give them a richer understanding of Node concepts and Node under the hood. That was also my intent with the "cooperative multitasking" link.